### PR TITLE
Improve, document and test AnnotationHelper

### PR DESCRIPTION
**Pull request by aykborstelmann**

### CODE REVIEW

File: OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/transformer/NepoOp.java [hunk 1]
```diff
@@ -7,19 +7,52 @@
...
+import de.fhg.iais.roberta.syntax.BlockType;
+import de.fhg.iais.roberta.syntax.BlockTypeContainer;
+import de.fhg.iais.roberta.syntax.BlocklyBlockProperties;
+import de.fhg.iais.roberta.syntax.BlocklyComment;
 import de.fhg.iais.roberta.syntax.lang.expr.Assoc;
+import de.fhg.iais.roberta.syntax.lang.expr.Expr;
 import de.fhg.iais.roberta.typecheck.BlocklyType;
 
+/**
+ * Annotate AST classes which are Expressions ({@link Expr}) and should be parsed with the new Annotation system with this.
```
Comment (rbudde) [inline OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/transformer/NepoOp.java:20]:
```suggestion
 * Annotate AST classes which are operators (the leaves in an expression {@link Expr}) and should be parsed with the Annotation system.
```

File: OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/transformer/NepoOp.java [hunk 2]
```diff
@@ -7,19 +7,52 @@
...
+ * This is primarly used to determine wheter this class should be parsed with new system, but also to define the {@link NepoPhrase#containerType()}.
```
Comment (rbudde) [inline OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/transformer/NepoOp.java:21]:
```suggestion
 * The annotation declares the basic properties of an operator (precedence, association), but also defines the {@link NepoPhrase#containerType()}.
```

File: OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/transformer/NepoPhrase.java [hunk 3]
```diff
@@ -7,14 +7,31 @@
...
+import de.fhg.iais.roberta.blockly.generated.Block;
+import de.fhg.iais.roberta.components.Category;
+import de.fhg.iais.roberta.syntax.BlockType;
+import de.fhg.iais.roberta.syntax.BlockTypeContainer;
+import de.fhg.iais.roberta.syntax.BlocklyBlockProperties;
+import de.fhg.iais.roberta.syntax.BlocklyComment;
 import de.fhg.iais.roberta.typecheck.BlocklyType;
 
+/**
+ * Annotate AST classes that should be parsed with the new Annotation system with this.
```
Comment (rbudde) [inline OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/transformer/NepoPhrase.java:19]:
```suggestion
 * Annotate AST classes that general phrases in the abstract syntax tree and should be parsed with the annotation system.
```

File: OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/transformer/NepoPhrase.java [hunk 4]
```diff
@@ -7,14 +7,31 @@
...
+ * This is primarly used to determine wheter this class should be parsed with new system, but also to define the {@link NepoPhrase#containerType()}.
```
Comment (rbudde) [inline OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/transformer/NepoPhrase.java:20]:
```suggestion
 * It is used to define the {@link NepoPhrase#containerType()}.
```

**Review by rbudde — COMMENTED** [review]
the reason to have default and required properties was to allow to annotate (probably bad designed) ast classes, that combine two independant concepts. This reduces the strictness of the annoation system, of course. If not needed, it makes sense to remove it. To me it is unclear, whether we'll find ast classes, that need sich properties.

File: OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/syntax/BlockTypeContainer.java [hunk 5]
```diff
@@ -117,4 +118,10 @@ public static BlockType getByBlocklyName(String blocklyName) {
         Assert.notNull(blockType, "blockly name is not found: " + blocklyName);
         return blockType;
     }
+
+    @VisibleForTesting
+    public static Map<String, BlockType> getBlockTypesByName() {
```
Comment (rbudde) [inline OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/syntax/BlockTypeContainer.java]: public? why not package visible?

Comment (rbudde) [reply OpenRobertaRobot/src/main/java/de/fhg/iais/roberta/syntax/BlockTypeContainer.java]: ok, was removed later. In general, the cost of checking valis annotations should occur only at server startup time and not for every parse of XML. Or?

**Comment by aykborstelmann** [comment]
I finished the implementation of Boolean, Double and Enum field values. Do you want to have a look over the new changes again ? @rbudde

**Pull request merged**